source/find_getmore_killcursors_commands.rst

```diff
@@ -365,18 +365,54 @@ More detailed information about the interaction of the **secondaryOk** with **OP
 Behavior of Limit, skip and batchSize
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The new **find** command has different semantics to the existing 3.0 and earlier **OP_QUERY** wire protocol message. The **limit** field is a hard limit on the total number of documents returned by the cursor no matter what **batchSize** is provided.
+The new **find** command has different semantics to the existing 3.0 and earlier
+**OP_QUERY** wire protocol message. The **limit** field is a hard limit on the
+total number of documents returned by the cursor no matter what **batchSize** is
+provided. This includes other limiting operations, such as the **$limit**
+aggregation pipeline stage. This differs from existing **OP_QUERY** behavior
+where there is no server-side concept of limit and where the driver **MUST**
+keep track of the limit on the client side and **MUST** send a
+**OP_KILL_CURSORS** wire protocol message when it limit is reached.
+
+When setting the **batchSize** on the **find** and **getMore** commands the
+value of **batchSize** **MUST** be based on the cursor limit calculations
+specified in the `CRUD`_ specification.
+
+Note that with 5.0, the server-side handling of cursors with a limit has
+changed. Before 5.0, some cursors were automatically closed when the limit was
+reached (e.g. when running **find** with **limit**), and the reply document did
+not include a cursor ID (i.e. ``cursor.id`` was ``0``). Starting with 5.0, all
+cursor-producing operations will return a cursor ID if the end of the batch
+being returned lines up with the limit on the cursor. In this case, drivers
+**MUST** ensure the cursor is closed on the server. To do this, drivers **MUST**
+use one of the options provided below:
+
+* send an additional **getMore** command with **batchSize** 1. Since the limit
+  has been reached, the **getMore** response is expected to contain neither a
+  cursor ID, nor any elements in the batch. If results are returned in the
+  **getMore** call, these **MUST** be discarded. If the result contains a cursor
+  ID, the driver **MUST NOT** attempt another **getMore**, but **MUST** use
+  **killCursors** as prescribed below.
```

Comment:
Apologies if this is a repeat question but I see this in the help ticket:
> Drivers that don't track the limit themselves will send out an additional getMore command for every cursor when the end of a batch lines up with the size of the cursor (e.g. count(cursor) == n * cursor.batchSize)

Do any drivers *not* track the limit? The existing test seems to suggest that all drivers do track the limit because of the updated getMore batchSize in the CommandStartedEvent. If so, would it be possible to add an extra command-monitoring test for >=5.0 which asserts that the driver sends a killCursors command?

Reply:
You are correct, drivers do track the limit, I believe this was an oversight when I wrote this.

As for the additional test, I had that in but @jyemin raised the point that drivers might not conform to this. For example, libmongoc sends an additional `getMore`, which means that adding a test that requires `killCursors` to be sent requires work in libmongoc, which we want to avoid here. I've only added guidance to the spec to make sure drivers check whether they handle this use-case correctly.

Once this PR is merged, I'll update the DRIVERS ticket with guidance for drivers what to do here, which is essentially:
* Check if the offending spec test fails on 5.0.0rc0
* Sync spec tests

If the spec test changed in this PR does not fail for a driver, it means that the driver does not send an additional command to exhaust or kill the cursor on the server. I'll leave it to each language team to decide the best course of action; it's not optimal to leave cursors open, but considering our busy workload it may not be possible to fix this at this time. These changes will be necessary once DRIVERS-1448 is done.